SSO UI MVP

rhysyngsun opened the pull request:
### What are the relevant tickets?

Part of https://github.com/mitodl/hq/issues/3473

### Description (What does it do?)
This updates our theme to align to the current designs.


### Screenshots (if appropriate):
![Screenshot 2024-03-20 at 14-19-14 MIT Open](https://github.com/mitodl/ol-keycloak/assets/28598/5e67a5bc-c984-494e-9127-56510ab49b4a)
![Screenshot 2024-03-20 at 14-50-29 MIT Open](https://github.com/mitodl/ol-keycloak/assets/28598/b924137d-a66d-426b-be4c-e902f53e5d85)
![Screenshot 2024-03-20 at 17-11-50 MIT Open](https://github.com/mitodl/ol-keycloak/assets/28598/b2f8912e-8ca4-4d68-b79b-2b4f9d5bb1e2)
![Screenshot 2024-03-20 at 14-58-58 MIT Open](https://github.com/mitodl/ol-keycloak/assets/28598/3e12377a-fee4-456b-8af9-004576a66e26)



### How can this be tested?
<!---
Please describe in detail how your changes have been tested.
Include details of your testing environment, any set-up required
(e.g. data entry required for validation) and the tests you ran to
see how your change affects other areas of the code, etc.
Please also include instructions for how your reviewer can validate your changes.
--->

##### Install the code changes:
- In `ol-keycloak/oltheme` and `ol-keycloak/ol-spi`, run `mvn clean install`
- Copy the JAR files in `target/` into your keycloak's `/opt/keycloak/providers` directory
- Start keycloak

##### Test functionality:

Through this testing, double check that everything matches the designs.

- Can register a new account w/ email confirmation
- Can login as an existing user
  - On the password page, there's a functional password visibility toggle.


### Additional Context

Note that this is a fairly large amount of changes which is primarily due to the fact that our original theme was based on the base keycloak theme in 23.0. Keycloak 24.0 came out about 2 weeks ago removing that and replacing it with a more modern theme (v2). This required me to gut quite a bit of our theme and reapply the customizations on top of the new v2 theme.

### Checklist:
- [ ] Update Keycloak to 24.0

collinpreston commented on ol-keycloak/ol-spi/src/main/java/forms/login/freemarker/OlFreeMarkerLoginFormsProvider.java:
```diff
@@ -36,16 +36,16 @@ public OlFreeMarkerLoginFormsProvider(KeycloakSession session) {
             UserModel user = context.getUser();
             if (context.getUser() !=  null) {
-                firstName = user.getFirstName();
+                attemptedName = user.getFirstName().concat(" ").concat(user.getLastName());
```
This line throws an error when the first name is null.  I'm not sure why that would occur in the real world, but I also don't know how we can be sure it never happens.

collinpreston reviewed (commented):
1. When hovering on the "Next" button on the username-password form, the cursor should change to the "hand".
2. When hovering on the "Send Reset Email" button on the forgot-password form, the cursor should change to the "hand".
3. When hovering on the "Next" button on the username form, the cursor should change to the "hand".
4. Mock-ups show the text "Sign In" being used on the username form instead of "Login".
5. The "Sign up" and "Forgot Password?" links on the username and username-password forms should be underlined to match the mock-ups.
6. When an email address is entered into the username-password form that does not match an existing Keycloak account, the error message displayed does not match the mock-ups.
7. When hovering on the "Submit" button on the register form, the cursor should change to the "hand".
8. The text on the register form submit button should be "Next" to match the mockups.
9. Should the text "©2024 Massachusetts Institute of Technology" be at the bottom of the pages?  I don't see it in the mock-ups.
10. The email-verification-sent page does not match the mock-up.
11. The formatting of the incorrect password error message on the username-password form does not match the mock-ups.
12. The _show password_ button is shown on the username-password form.

rhysyngsun commented:
@collinpreston I addressed most of your feedback, there are some things from the mockups that aren't feasible with theme changes or I left for other reasons, more details below:

>     6. When an email address is entered into the username-password form that does not match an existing Keycloak account, the error message displayed does not match the mock-ups.

I updated this to be as close as I can to the mock up, but I can't put a link in that message or dynamic information like the email address in there so I adapted the wording as best I can. I made Steve aware there are issues like this.

>     9. Should the text "©2024 Massachusetts Institute of Technology" be at the bottom of the pages?  I don't see it in the mock-ups.
> 

Asked on Slack, @pdpinch said we some kind of minimal footer, including copyright, at some point despite it not being in the mocks, so I'll leave this in.

>     10. The email-verification-sent page does not match the mock-up.

I looked into this, it can't be accomplished by templates, I didn't look into the feasibility but it would require a new Java customization if it can be done, that's out of scope for what I'm doing here. It's been asked a few times in various community forums and received no response as to how to even attempt it.

>     11. The formatting of the incorrect password error message on the username-password form does not match the mock-ups.

I updated this to be as close as I can to the mock up, but same as the email fields I can't put a link in that message so I carried over the "select Forgot Password below" wording. There's likely going to be more adjustments to the UX after I merge anyway, so this is probably fine for now.

>     12. The _show password_ button is shown on the username-password form.

This was intentional, there was a separate figma file from steve that noted this as a nice to have (#3 on second row): https://www.figma.com/file/M7raxNajixAyAB2MNfAB6l/MIT-Canvas?type=design&node-id=456-11513&mode=design

pdpinch commented:
Thanks for enumerating those. I think they are fine for now (and maybe fine forever), so we can move ahead. I'm eager to test this on RC.

collinpreston reviewed (changes requested):
1. For an existing user with a password, the username-password page displays "You already have a login with:" even if that user has no identity provider linked.  There is nothing displayed below the "You already have a login with:" text.
2. The MIT Open logo at the top of each page is only clickable on the left side of the logo.

rhysyngsun commented:
@collinpreston all set
